qiskit/algorithms/optimizers/spsa.py

```diff
@@ -49,195 +53,335 @@ class SPSA(Optimizer):
+
+        # get iterator
+        eta = self.learning_rate()
+        eps = self.perturbation()
+
+        # prepare some initials
+        x = np.asarray(initial_point)
+
+        self._nfev = 0
+
+        # if blocking is enabled we need to keep track of the function values
+        if self.blocking:
+            fx = loss(x)
+
+            self._nfev += 1
+            if self.allowed_increase is None:
+                self.allowed_increase = 2 * self.estimate_stddev(loss, x)
+
+        logger.info('=' * 30)
+        logger.info('Starting SPSA optimization')
+        start = time()
+
+        # keep track of the last few steps to return their average
+        last_steps = deque([x])
+
+        for k in range(1, self.maxiter + 1):
+            iteration_start = time()
+            # compute update
+            update = self._compute_update(loss, x, k, next(eps))
+
+            # trust region
+            if self.trust_region:
+                norm = np.linalg.norm(update)
+                if norm > 1:  # stop from dividing by 0
+                    update = update / norm
+
+            # compute next parameter value
+            update = update * next(eta)
+            x_next = x - update
+
+            # blocking
+            if self.blocking:
+                fx_next = loss(x_next)
+
+                self._nfev += 1
+                if fx + self.allowed_increase <= fx_next:  # accept only if loss improved
+                    if self.callback is not None:
+                        self.callback(self._nfev,  # number of function evals
+                                      x_next,  # next parameters
+                                      fx_next,  # loss at next parameters
+                                      np.linalg.norm(update),  # size of the update step
+                                      False)  # not accepted
+
+                    logger.info('Iteration %s/%s rejected in %s.',
+                                k, self.maxiter + 1, time() - iteration_start)
+                    continue
+                fx = fx_next
+
+            logger.info('Iteration %s/%s done in %s.',
+                        k, self.maxiter + 1, time() - iteration_start)
+
+            if self.callback is not None:
+                self.callback(self._nfev,  # number of function evals
+                              x_next,  # next parameters
+                              fx_next,  # loss at next parameters
+                              np.linalg.norm(update),  # size of the update step
+                              True)  # accepted
+
+            # update parameters
+            x = x_next
+
+            # update the list of the last ``last_avg`` parameters
+            if self.last_avg > 1:
+                last_steps.append(x_next)
+                if len(last_steps) > self.last_avg:
+                    last_steps.popleft()
+
+            # check termination
+            if np.linalg.norm(update) < self.tolerance:
```

Comment:
why break if tolerance is true for a single iter given that everything is stochastic?  The next iter could be outside this tolerance.  Does it not make sense to track this as an ensemble average?

Reply:
good point... maybe a tolerance is not a great idea for SPSA. I'll remove it until we come up with something better 👍🏻